### PR TITLE
Update e2e tests to use convert_model

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

- Update torch examples, added argument `--export-model-path` instead of `--to-onnx` that select output type by suffix '.xml' or '.onnx'.
- Update tests to use Command to run examples.
- Remove tests of q_dq export path.
- Remove code to make html report, now metrics dumps in `results.csv`.
- Added "build url" in `results.csv`.
- Actualize reference metrics. 
- Move normalization of input to accuracy cheker configs.
- Use `target_ov` and `target_pt` to define reference metrics for backends.  
- Skip test `unet_mapillary_int8` and `unet_mapillary_magnitude_sparsity_int8` models by 123448.
- Skip `resnet18_imagenet_binarization_dorefa` model by 22543.

Updates in CI:
- Update script to report `final e2e_result.html` for e2e tests.
- Now if any test falls trigger_job will be fails.

After merge requires update ci-pipelines.

### Related tickets

117885

### Tests

tests/torch/test_sota_checkpoints.py

### TODO
- update metrics after merge   https://github.com/openvinotoolkit/nncf/pull/2227
- check metrics after https://github.com/openvinotoolkit/nncf/pull/2211

### CODE REVIEW

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [inline tests/torch/test_sota_checkpoints.py:95]: Perhaps this could also be encapsulated within a class-level fixture instead of being a global var.

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [inline tests/torch/test_sota_checkpoints.py:139]: I think `TEST_ROOT` is already a path.

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [inline tests/torch/test_sota_checkpoints.py:133]: The default value should probably be `None` or `False`

**Review by vshampor — DISMISSED** [review]
Good to see usage of dataclasses, test fixtures and type hints - don't forget to take opportunities to set up type hints even if they weren't therer before.

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py:145]: Please provide example of bad params about you tell.  
It's not clear how adding `NotImplementedError` can help with tests with same ID.
And how you can get same ID for multiple tests?

File: examples/torch/common/argparser.py
Comment (AlexanderDokuchaev) [reply examples/torch/common/argparser.py]: Valid, change argument to `--export-model-path` insated of `--to-ir` or `--to-onnx`
```
- Use `--export-model-path` to specify the path to export the model in OpenVINO or ONNX format by using the .xml or .onnx suffix, respectively.
```

File: examples/torch/common/argparser.py
Comment (AlexanderDokuchaev) [reply examples/torch/common/argparser.py]: update

File: examples/torch/common/export.py
Comment (AlexanderDokuchaev) [reply examples/torch/common/export.py:23]: Thanks, fixed

File: examples/torch/common/export.py
Comment (AlexanderDokuchaev) [reply examples/torch/common/export.py]: Yes, fixed

File: examples/torch/common/export.py
Comment (AlexanderDokuchaev) [reply examples/torch/common/export.py]: moved import in function

File: examples/torch/object_detection/configs/ssd512_vgg_voc_magnitude_sparsity_int8.json
Comment (AlexanderDokuchaev) [reply examples/torch/object_detection/configs/ssd512_vgg_voc_magnitude_sparsity_int8.json]: https://github.com/openvinotoolkit/nncf/pull/2152#discussion_r1370900477 (about "--export-model-path") 
Even so, `export_via_onnx` is still requited because some models can not be converted without onnx.

File: tests/torch/sota_checkpoints_eval.json
Comment (AlexanderDokuchaev) [reply tests/torch/sota_checkpoints_eval.json:507]: On practic for OV it's not equal, and more for PT that tested on GPU we get different metric build by build. 
So separated `target` and `target_diff` for PT and OV is good solution for this test.

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py:147]: If you worry about that you can add test case in `sota_checkpoints_eval.json` with not unique model name, it's possible, added check in `read_reference_file`.

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: fixed

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: But this name was approved by your, https://github.com/openvinotoolkit/nncf/pull/1017.
But ok, conftest.py update and ci-pipelines too .

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py:254]: Look on `results.csv` file, that contains description of errors in new "Status" column. 
Html report available on Jenkins.

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: This name was approved by your in https://github.com/openvinotoolkit/nncf/pull/445
But ok

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: And this was approved by your https://github.com/openvinotoolkit/nncf/pull/445
Please be more consistent.

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: fixed

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: yes

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: Fixed

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: And what a problem to use setup_class? And how is the question "How to pass an argument to setup_class?" relate to this pr?

File: tests/torch/sota_checkpoints_eval.json
Comment (AlexanderDokuchaev) [reply tests/torch/sota_checkpoints_eval.json:525]: Test skipped until issue will not be fixed or found solution to get stable results.
We have no free resources to run long tests that will be guaranty falls.

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py:95]: There is any problem to use with global var for this case?

**Comment by AlexanderDokuchaev** [comment]
view/all/job/ci-maintenance/job/pt_e2e_tests/job/NNCF/job/nightly/job/TriggerBetta/55
view/all/job/ci-maintenance/job/pt_e2e_tests/job/NNCF/job/nightly/job/TriggerBetta/57 (with fixed reference)

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py:147]: Looks like you want to be handling the case when the idfn can accept something other than `EvalRunParamsStruct`. I think you want to say that this situation is invalid, and so you return `None` as an ID. This leads to no errors in runtime, the tests will still be run. Instead you should make the situation fail even before the tests are started, so that no one mistakenly adds a non-`EvalRunParamsStruct` object to the list of parameters that this `idfn` is related to.

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py]: I am expecting you to do better at this point of time than junior devs and contingent workers.

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py]: You should have read the SO answer before responding in such a manner. I guess I'll have to point out the relevant paragraph to you:
>you are trying to mix two independent testing styles that py.test supports: the classical unit testing and pytest's fixtures.
>What I suggest is not to mix them and instead simply define a class scoped fixture like this:

So you are violating consistency in our test code by using `setup_class`.

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py:95]: Global state is bad, period, if it can be avoided. Specifically in the situation of our E2E tests we are launching these in a separate python process. As long as it is true, I suppose the global vars can functionally be used and do not pose visible problems. However, if this situation changes in the future, e.g. we will be launching the E2E tests along with some other tests, or add a test class to the E2E testing suite that would try to reuse this global state, we might be ending with hard-to-discover subtle bugs.

Your state actually belongs to the E2E testing specifically. If you encapsulate E2E test code in a speicific class (it seems that this is done already), then this state should also be associated with the class. With fixtures it seems to be done using a class-scoped fixture that would be returning an object reference that is shared across all tests in a class, and you can use this object to store test results as tests execute. See https://stackoverflow.com/questions/28288385/pytest-how-to-pass-an-argument-to-setup-class again - if I'm not mistaken, the `A_helper` object will only be created once per class, accessible in every test case and can therefore serve as a storage for class-scope state such as the list of the test results.

Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py:95]: ` Specifically in the situation of our E2E tests we are launching these in a separate python process. As long as it is true, I suppose the global vars can functionally be used and do not pose visible problems. ` 

Ok i will keep as it.

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: Do you can answers on my question?

> So you are violating consistency in our test code by using setup_class.

Is it exist in contribution or style guide in NNCF?

Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py]: I have posted a relevant link, even copied the relevant paragraph, and provided a corollary argument, all in [this comment](https://github.com/openvinotoolkit/nncf/pull/2152#discussion_r1371451401). This is not the first time you ask me to give you an answer that I already gave in a previous comment. If you have problems with reading comprehension in English, reach out to me offline and I'll translate to Russian for you.

Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py]: > Is it exist in contribution or style guide in NNCF?
I can add it, no problem.

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py:95]: > `Specifically in the situation of our E2E tests we are launching these in a separate python process. As long as it is true, I suppose the global vars can functionally be used and do not pose visible problems.`
> 
> Ok i will keep as it.

As a good developer, you should be preparing for the future, and implementing best practices in code even if they don't seem to have an immediate effect, especially if you are pointed towards them explicitly. So I, as a maintainer, want you to do so and won't approve the PR otherwise.

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py]: See #2220

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py:153]: To be clear, your comment confused me a bit, because as I know:

- pytest does not allow duplicate test ids by automatically adds an index to duplicate ids from idfn;
- if `idfn` returns None, will be used the default pytest converter to test id.

I am curious to know how it is possible to achieve the behavior you described.

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: was fixed

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: Ok, please look on 
https://github.com/openvinotoolkit/nncf/blob/6130fca6dc199ea0b0a8a4dfa785e0da9217ca66/tests/torch/test_sota_checkpoints.py#L257-L270

Is it correct way to use fixture to collect data?

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: I am still asking you because your comments do not contain any answers on my quastions.

From your pr is it clear, thanks.

File: tests/torch/sota_checkpoints_eval.json
Comment (AlexanderDokuchaev) [reply tests/torch/sota_checkpoints_eval.json]: Marked as xfail, will looks like
image

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py]: I didn't know that - I thought that returning `None` would generate a test ID of `"None"`, I've been seeing some of the test IDs with `"None"` but I guess they ended up in the test ID due to some other mechanism. So your current code will indeed pose no problems with duplicate IDs.

However, do you really want to do the `isinstance` check and allow non-`EvalRunParamsStruct` objects to be parameters for your test case? If a non-`EvalRunParamsStruct` is passed as a parameter, then this is most likely an error on part of the developer who edited the params list. Maybe it would be better to set up a type hint in the `idfn` signature and let the ID generation fail with an "AttributeError" if someone mistakenly puts a wrong object into test case params.

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py:254]: Ok, so it seems that the visualization code is now run from the CI pipeline. I don't really like this since this takes the visualization out of our hands and we cannot improve it if necessary. The popular vote was for the new visualization format, so I won't block on this.

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py]: This is how it's supposed to be done, yes. Details could be adjusted if necessary. It worked well to date, anyway.

File: examples/torch/common/argparser.py
Comment (vshampor) [reply examples/torch/common/argparser.py]: `--to-onnx` and `--to-ir` are shorter, though, easier to type, and you won't be breaking existing code that relies on `--to-onnx` if you keep `--to-onnx` and just add `--to-ir`.

File: tests/torch/sota_checkpoints_eval.json
Comment (vshampor) [reply tests/torch/sota_checkpoints_eval.json:507]: Metric for PT is expected to differ from run to run, but also expected to be minor, stay in a given range (which is why we set up a threshold of `0.1`) and be distributed around a given value (which we call `target`). If OV runtime does not reproduce this PT value, then there is a bug somewhere, because OV is expected to do inference functionally the same as the framework (with only minor discrepancies). So the target for OV is defined exclusively by the target for PT (it is equal to it) in the normal case, and the normal case should only specify `"target_ov"` if there is something unfixable about OV metrics differing from PT metrics. Right now only SSD models would be the candidates for specifying `"target_ov"`, other models have OV metrics conformant to the PT metric and it should stay so.

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py]: test_id collect `None` if parameter is `None`, it is expected.

It's default template of idfn that provided in https://docs.pytest.org/en/7.4.x/example/parametrize.html#different-options-for-test-ids. And it's used in other NNCF tests.
If somebody will pass to incorrect data to fixture, test will fall anyway. Not `EvalRunParamsStruct` parameters can be used for idfn, and idfn can be extended to types.

File: tests/torch/sota_checkpoints_eval.json
Comment (AlexanderDokuchaev) [reply tests/torch/sota_checkpoints_eval.json:507]: Ofc, it can be easly done by `target_pt = sample_dict.get("target_pt", sample_dict["target_ov"]`.
But differ will be only that reference will be less on several lines, and there is no other improvements.
OV and PT_CUDA metrics is not equal in general case, so i have no see any problem to have two targets in reference file.

So looks like 'target_ov' should be prioritized metrics for test because:
- OV provides stable results 
- we can expect 0.0 in 'Target Diff' for all models 
- not clear which metrics value should be used for target_pt in case of unstable result (in case where i update metric for pt, i used mean for several runs).
- and final argument: it's e2e test and the final step is validation OV model as target inference backend.

So, i can keep current state or i can use `target_ov` as default value for `target_pt` and add `target_pt` only for ssd models?


@AlexKoff88  @alexsu52  what do think about it?

File: tests/torch/test_sota_checkpoints.py
Comment (vshampor) [reply tests/torch/test_sota_checkpoints.py]: Looking out for more types than `EvalRunParamStruct` is probably overengineering here. But now that you've explained your intent, the solution makes sense.

File: tests/torch/sota_checkpoints_eval.json
Comment (alexsu52) [reply tests/torch/sota_checkpoints_eval.json:507]: Honestly, I don't see any difference in these solutions, both apply. Historically, we always use only target metric from framework to calculate accuracy drop. @KodiaqQ what is your opinion?

Comment (KodiaqQ) [reply tests/torch/sota_checkpoints_eval.json:507]: In the ideal scenario, we should have only one column that we should rely on - `reference fp32`, and calculate a 1% `absolute` drop (if applicable) for all cases: OV, PT, TF, no matter. In this scenario, it's valid to extend the drop threshold in individual cases (like SSD), but it should be calculated on the fly in general.
This is my opinion, and the final decision should be made by a majority vote.

File: examples/torch/common/export.py
Comment (KodiaqQ) [inline examples/torch/common/export.py:62]: Why is compression to the FP16 precision was disabled?

File: tests/torch/test_sota_checkpoints.py
Comment (KodiaqQ) [inline tests/torch/test_sota_checkpoints.py:34]: What is the `minimum` difference? Why is the target for PyTorch and OpenVINO 10 times differ?

File: examples/torch/common/export.py
Comment (AlexanderDokuchaev) [reply examples/torch/common/export.py]: Removed it, ci-maintenance/job/pt_e2e_tests/job/NNCF/job/nightly/job/TriggerBetta/75/

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py:34]: > What is the minimum difference?
https://github.com/openvinotoolkit/nncf/blob/08cd0f7de2de04d0cdaf1c8b326ad96323c342af/tests/torch/test_sota_checkpoints.py#L314

>Why is the target for PyTorch and OpenVINO 10 times differ?

Because metrics measured on PT are unstable.

Comment (KodiaqQ) [reply tests/torch/test_sota_checkpoints.py:34]: > Because metrics measured on PT are unstable.

In my opinion, it's not the reason to roughly extend the threshold for all models 10 times. Also, there was no such threshold in the old version of the test.
In any case, we should decide this with the thoughts of @alexsu52 and @vshampor.

File: examples/torch/common/argparser.py
Comment (AlexanderDokuchaev) [reply examples/torch/common/argparser.py]: The comment has been reviewed and no further action will be taken. 
@MaximProshin

File: tests/torch/test_sota_checkpoints.py
Comment (AlexanderDokuchaev) [reply tests/torch/test_sota_checkpoints.py:34]: > In my opinion, it's not the reason to roughly extend the threshold for all models 10 times. 

I did not extend thresholds for all models, by default thresholds was +-0.1, now this thresholds keep for PT and reduced for OV.


> Also, there was no such threshold in the old version of the test.
It was.
 https://github.com/openvinotoolkit/nncf/blob/08cd0f7de2de04d0cdaf1c8b326ad96323c342af/tests/torch/test_sota_checkpoints.py#L304-L314

**Review by KodiaqQ — APPROVED** [review]

**Review by alexsu52 — APPROVED** [review]
Great stuff!!!

**Pull request merged**
